### PR TITLE
Do not allow to delete volume from brokers

### PR DESCRIPTION
**Pull request by bartam1**
| Q               | A      |
| --------------- | ------ |
| Bug fix?        | yes |
| New feature?    | no |
| API breaks?     | no|
| Deprecations?   | no |
| Related tickets | #739  |
| License         | Apache 2.0 |

### What's in this PR?
- There is kafkaCluster validation webhook that checks is there a broker storage removal. When there is then the validation will be failed.
- There is another protection for this problem in this PR. When there is broker storage removal in the modified kafkaCluster CR, the koperator will not remove it in the reconcile flow from the broker "log.dirs".

### Why?
Currently nothing prohibits the user to remove broker storage entry from KafkaCluster custom resource. This will trigger koperator to remove the storage from broker configuration. This should not be allowed because it can lead to data loss

### Checklist
- [x] Implementation tested
- [x] Error handling code meets the [guideline](https://github.com/banzaicloud/developer-guide/blob/master/docs/coding-style/error-handling-guide.md)
- [x] Logging code meets the guideline

### CODE REVIEW

**Review by panyuenlau — DISMISSED** [review]
btw `mounth` is a typo and it should be `mount`

File: pkg/webhook/kafkacluster_validator.go
Comment (bartam1) [reply pkg/webhook/kafkacluster_validator.go:77]: Im not sure it would be faster. hash function also need time. searching in an array sometimes faster when the array is not so big. There is more chance for cache hit because locality. I will make a benchmark and I will write the result here then. But Im almost sure if we don't have 10000 broker the iteration is faster then the map.

File: pkg/resources/kafka/configmap.go
Comment (bartam1) [reply pkg/resources/kafka/configmap.go:123]: Good catch. It is only a log.Error :)

Comment (bartam1) [reply pkg/resources/kafka/configmap.go:123]: It handles because (var brokerConfigMapOld v1.ConfigMap) this is not nil. It contains Data map[string]string. 
brokerConfig := configMap.Data[kafkautils.ConfigPropertyName]  here brokerConfig will be empty string, and return will be empty slice

Comment (bartam1) [reply pkg/resources/kafka/configmap.go:123]: Im going to put a protection in it anyway.

File: pkg/webhook/kafkacluster_validator.go
Comment (bartam1) [reply pkg/webhook/kafkacluster_validator.go]: I have made tests:
With 10 brokers:
BenchmarkCheckBrokerStorageRemoval-12             336151             17843 ns/op            7709 B/op        181 allocs/op
BenchmarkCheckBrokerStorageRemovalMap-12          315955             20312 ns/op            7709 B/op        181 allocs/op
with 200 brokers:
BenchmarkCheckBrokerStorageRemoval-12               1857           3301253 ns/op         1534065 B/op      36213 allocs/op
BenchmarkCheckBrokerStorageRemovalMap-12            1784           3638668 ns/op         1566985 B/op      36224 allocs/op
With 1000 brokers:
BenchmarkCheckBrokerStorageRemoval-12                235          24051552 ns/op         7701831 B/op     181791 allocs/op
BenchmarkCheckBrokerStorageRemovalMap-12             273          21840247 ns/op         7959142 B/op     181814 allocs/op
With 20000 brokers:
BenchmarkCheckBrokerStorageRemoval-12                  6        1067613391 ns/op        155720769 B/op   3639313 allocs/op
BenchmarkCheckBrokerStorageRemovalMap-12              14         408982641 ns/op        158881898 B/op   3639976 allocs/op

Because almost impossible to have 1000broker cluster, I would stay with the for loop.

Comment (bartam1) [reply pkg/webhook/kafkacluster_validator.go]: The important field here is the ns/op

Comment (bartam1) [reply pkg/webhook/kafkacluster_validator.go]: I added benchmark tests, and you and the team can also check. Before the merge those tests will be removed

File: pkg/webhook/kafkacluster_validator.go
Comment (bartam1) [reply pkg/webhook/kafkacluster_validator.go]: Probably the name here is ambiguous. If I can found the old storage in the new one then the isStorageFound will be true. But If I cant, then it will be false, and I will reject.

File: pkg/webhook/kafkacluster_validator_test.go
Comment (pregnor) [inline pkg/webhook/kafkacluster_validator_test.go:174]: **Note**: unnecessary comment? (unless it is to understand the case more easily)

File: pkg/webhook/kafkacluster_validator.go
Comment (pregnor) [inline pkg/webhook/kafkacluster_validator.go:78]: **Question**: Why are we using `nil` here to signal allowed case instead of using an allowed response?

I see no distinct meaning of `nil` vs `allowed` here, but we use two different states for the same behavior.

If we were to use the `allowed` here instead of `nil` we could also directly return the result of this function from the validation handler for now.

Note: this affects the table testing logic as well (that is the place I got the idea from).

File: charts/kafka-operator/templates/operator-deployment-with-webhook.yaml
Comment (bartam1) [reply charts/kafka-operator/templates/operator-deployment-with-webhook.yaml:61]: Currently koperator is architected this way: https://github.com/banzaicloud/koperator/blob/b9fe4efd6c28ad8292af22f2c48197388d736754/pkg/webhook/request.go#L44

File: pkg/webhook/request.go
Comment (bartam1) [reply pkg/webhook/request.go:76]: Reconcile will not run on cluster registry managed ones, and the copy will not replicated back if I know good.

File: pkg/webhook/kafkacluster_validator_test.go
Comment (bartam1) [reply pkg/webhook/kafkacluster_validator_test.go:174]: It can be removed, but you can see also more easily that this one has been removed from the new Config. So this validation should be rejected

File: pkg/webhook/kafkacluster_validator.go
Comment (bartam1) [reply pkg/webhook/kafkacluster_validator.go:78]: True. I just followed what is at the kafkatopic validation.  (https://github.com/banzaicloud/koperator/blob/b9fe4efd6c28ad8292af22f2c48197388d736754/pkg/webhook/topic_validator.go#L103)
I think it would be nicer if I don't return here nil.
Im going to change it. Thanks

File: charts/kafka-operator/templates/operator-deployment-with-webhook.yaml
Comment (pregnor) [reply charts/kafka-operator/templates/operator-deployment-with-webhook.yaml:61]: Oh I see, my bad, thanks.

File: pkg/webhook/kafkacluster_validator.go
Comment (pregnor) [reply pkg/webhook/kafkacluster_validator.go:69]: Oh my bad on the implementation then, it should be `if isExistingID && !isExistingMountPath`.

I'm fine with the current solution as well.

**Review by pregnor — DISMISSED** [review]

File: pkg/webhook/kafkacluster_validator_test.go
Comment (Kuvesz) [reply pkg/webhook/kafkacluster_validator_test.go:266]: If we wish to show that should we not have a comment indicating that instead? Like instead of this you could write in the comment that the removed MountPath is this.

**Review by pregnor — APPROVED** [review]

File: pkg/webhook/request.go
Comment (Kuvesz) [inline pkg/webhook/request.go:42]: If we have AdmissionReview printed in every case why not just leave it here? Is it only because of the kafkaCluster case having the if in the beginning?

**Review by Kuvesz — DISMISSED** [review]
LGTM, left some very minor comments.

File: pkg/webhook/request.go
Comment (bartam1) [reply pkg/webhook/request.go:42]: Because if it would be there, then in every case this info would be written in the operator log, but now there is only one exception here: https://github.com/banzaicloud/koperator/blob/3c935f3219cd49cf12f3d8294b6810946ec75219/pkg/webhook/request.go#L60 . This means when the operator change the kafkacluster, we does not validate and write "AdmissionReview", because it happens many times. It pollutes the log unnecessarily.

**Review by Kuvesz — APPROVED** [review]

**Review by Kuvesz — APPROVED** [review]

**Review by Kuvesz — DISMISSED** [review]

File: pkg/webhook/request.go
Comment (bartam1) [reply pkg/webhook/request.go:60]: Thank you! I will implement this.

**Review by stoader — CHANGES_REQUESTED** [review]

File: pkg/webhook/kafkacluster_validator.go
Comment (stoader) [inline pkg/webhook/kafkacluster_validator.go:47]: The logic implemented in this function doesn't take into account brokers for which a downscale is in progress. The volumes used by that broker should not be allowed to be removed until the broker is gracefully downscaled. So brokers from the `status` field that are in either pending or running downscale needs to be taken into account as well beside the list of brokers in the spec.

**Review by stoader — CHANGES_REQUESTED** [review]

File: pkg/resources/kafka/configmap.go
Comment (stoader) [inline pkg/resources/kafka/configmap.go:238]: Why don't you use `properties.NewProperties()` utility to parse `brokerConfig` and extract the log dir configuration using its `Get` method?

**Review by stoader — CHANGES_REQUESTED** [review]

File: charts/kafka-operator/templates/operator-deployment-with-webhook.yaml
Comment (pregnor) [reply charts/kafka-operator/templates/operator-deployment-with-webhook.yaml:61]: I agree with Sebastian.

The latest version of Kubebuilder [advocates](https://book.kubebuilder.io/reference/webhook-for-core-types.html) for using Sebastian's proposed solution with the latest controller-runtime version using [admission 
webhook handlers](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/webhook/admission#Handler) for implementing validating webhooks ([official example](https://github.com/kubernetes-sigs/controller-runtime/blob/master/examples/builtins/validatingwebhook.go)).

What we may also do here alternatively if we want to unblock this PR due to having it open for an extended amount of time is to let it in using the current implementation and have a refactoring issue for this later on, but I heavily agree with Sebastian's point that if we ever want to modernize the implementation to the latest version (which we should IMO), the best time to start is right now when we have to touch it anyway.

(Currently irrelevant sidenote: For CRDs with simple localized validation - that can be done without external operational information - there is also a possibility to implement the [`admission.Validator`](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.12.1/pkg/webhook/admission#Validator) interface and [have the handler and the webhook generated from it](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.12.1/pkg/webhook/admission/validator.go#L38) with [`admission.ValidatingWebhookFor()`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/webhook/admission#ValidatingWebhookFor) ([kubebuilder example](https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_webhook.go)).
This of course could not work here due to the requirement to check whether the update request came from the operator or not.)

File: pkg/resources/kafka/configmap.go
Comment (bartam1) [reply pkg/resources/kafka/configmap.go:157]: This solution checks the mountPath regardless the index. 
Back to our talk, we need to consider also that case when the user only change the size of the storage. In this case, the old and new storageConfig will be different but the mountPath value remains the same.

**Review by pregnor — DISMISSED** [review]
LGTM

File: pkg/webhook/kafkacluster_validator_test.go
Comment (stoader) [reply pkg/webhook/kafkacluster_validator_test.go:266]: I agree with @Kuves

File: pkg/webhook/request.go
Comment (stoader) [inline pkg/webhook/request.go:46]: Instead of `l.Info("AdmissionReview")` you could initialize the logger as  
```
l := log.WithName("AdmissionReview").WithValues("kind", req.Kind, "namespace", req.Namespace, "name", req.Name, "uid", req.UID,
 		"operation", req.Operation, "user info", req.UserInfo)
```

File: pkg/webhook/request.go
Comment (stoader) [inline pkg/webhook/request.go:61]: Inject the namespace and service account instead of always reading from env var

**Review by stoader — CHANGES_REQUESTED** [review]

**Review by pregnor — APPROVED** [review]

**Review by stoader — APPROVED** [review]

**Review by Kuvesz — APPROVED** [review]

**Pull request merged**
